openedx/core/djangoapps/content_tagging/api.py

```diff
@@ -175,6 +190,118 @@ def tag_content_object(
+            Q(tag__isnull=False, tag__taxonomy__isnull=False),
+        ).select_related("tag__taxonomy")
+
+    def _group_object_tags_by_objectid_taxonomy(
+        all_object_tags: list[ObjectTag]
+    ) -> tuple[ObjectTagByObjectIdDict, TaxonomyDict]:
+        """
+        Returns a tuple with a dictionary of grouped object tags for all blocks and a dictionary of taxonomies.
+        """
+
+        groupedObjectTags: ObjectTagByObjectIdDict = {}
+        taxonomies: TaxonomyDict = {}
+
+        for object_id, block_tags in groupby(all_object_tags, lambda x: x.object_id):
+            groupedObjectTags[object_id] = {}
+            for taxonomy_id, taxonomy_tags in groupby(block_tags, lambda x: x.tag.taxonomy_id):
+                object_tags_list = list(taxonomy_tags)
+                groupedObjectTags[object_id][taxonomy_id] = object_tags_list
+
+                if taxonomy_id not in taxonomies:
+                    taxonomies[taxonomy_id] = object_tags_list[0].tag.taxonomy
+
+        return groupedObjectTags, taxonomies
+
+    def _build_object_tree_with_objecttags(
+        content_key: ContentKey,
+        include_children: bool,
+        objectTagCache: ObjectTagByObjectIdDict,
+        store
+    ) -> TaggedContent:
+        """
+        Returns the object with the tags associated with it. If include_children is True, then it will also include
+        the children of the object and their tags.
+        """
+        if isinstance(content_key, CourseKey):
+            xblock = store.get_course(content_key)
+        elif isinstance(content_key, UsageKey):
+            xblock = store.get_item(content_key)
+        else:
+            raise NotImplementedError(f"Invalid content_key: {type(content_key)} -> {content_key}")
+
+        tagged_xblock = TaggedContent(
+            xblock=xblock,
+            object_tags=objectTagCache.get(str(content_key), {}),
+            children=None,
+        )
+
+        if not include_children:
+            return tagged_xblock
+
+        blocks = [tagged_xblock]
+
+        while blocks:
+            block = blocks.pop()
+            block.children = []
+
+            if block.xblock.has_children:
+                for child_id in block.xblock.children:
+                    child = TaggedContent(
+                        xblock=store.get_item(child_id),
+                        object_tags=objectTagCache.get(str(child_id), {}),
+                        children=None,
+                    )
+                    block.children.append(child)
+
+                    blocks.append(child)
+
+        return tagged_xblock
+
+    store = modulestore()
+
+    all_blocks_tag_records = list(_get_object_tags(content_key, include_children))
+    objectTagCache, taxonomies = _group_object_tags_by_objectid_taxonomy(all_blocks_tag_records)
+
+    return _build_object_tree_with_objecttags(content_key, include_children, objectTagCache, store), taxonomies
```

Comment:
**I think the code for this function is pretty good, but you could organize it better:**

(1) You should split `get_object_tree_with_objecttags` into two functions:
    a. `get_all_object_tags` (get all object tags in a course, grouped by object ID, as well as a list of all the taxonomies that are used in the course) - this is the first half of your function, i.e. `_get_object_tags` and `_group_object_tags_by_objectid_taxonomy`. Notice that these functions are not used recursively and each is only called once. I think if you remove the parts I pointed out that we don't need, you can flatten this into one small function rather than two nested functions. If you want to keep it as two functions though, that's fine - just make them two top-level private functions.
   b. `_build_object_tree_with_objecttags` a separate recursive function.

(2) I don't think `_build_object_tree_with_objecttags` should be part of the python API for `content_tagging`. Because getting a tree of the XBlocks in a course is not really something the content tagging API should provide. I would move `_build_object_tree_with_objecttags` (as well as the `TaggedContent` type definition) to a separate private file, like `rest_api/v1/tag_export_helpers.py`

But `get_all_object_tags` can/should be part of the public API, because I don't think there is any other way to achieve this using the API today.

By separating these you can also write unit tests for them individually, if you like.

Reply:
Done: 7a2874246077446d49d635c8643907948e65e63e